src/cls/rbd/cls_rbd.cc

```diff
@@ -5975,108 +5982,1206 @@ int mirror_remote_namespace_set(cls_method_context_t hctx, bufferlist *in,
+  /*
+  if (group.state != cls::rbd::MIRROR_GROUP_STATE_DISABLING) {
+    return -EBUSY;
   }
+  */
 
-  int r = mirror::peer_ping(hctx, site_name, mirror_uuid);
+  r = cls_cxx_map_remove_key(hctx, group_key(group_id));
   if (r < 0) {
+    CLS_ERR("error removing mirrored group '%s': %s", group_id.c_str(),
+            cpp_strerror(r).c_str());
     return r;
   }
 
-  return 0;
-}
-
-/**
- * Input:
- * none
- *
- * Output:
- * @param std::vector<cls::rbd::MirrorPeer>: collection of peers
- * @returns 0 on success, negative error code on failure
- */
-int mirror_peer_list(cls_method_context_t hctx, bufferlist *in,
-                     bufferlist *out) {
-  std::vector<cls::rbd::MirrorPeer> peers;
-  int r = mirror::read_peers(hctx, &peers);
+  r = cls_cxx_map_remove_key(hctx, group_global_key(group.global_group_id));
   if (r < 0 && r != -ENOENT) {
+    CLS_ERR("error removing global id for group '%s': %s", group_id.c_str(),
+           cpp_strerror(r).c_str());
+    return r;
+  }
+
+  r = group_status_remove(hctx, group.global_group_id);
+  if (r < 0) {
+    CLS_ERR("group status remove error for group '%s': '%s'", group_id.c_str(),
+            cpp_strerror(r).c_str());
     return r;
   }
 
-  encode(peers, *out);
   return 0;
 }
 
-/**
- * Input:
- * @param mirror_peer (cls::rbd::MirrorPeer)
+int group_status_set(cls_method_context_t hctx, const string &global_group_id,
+		     const cls::rbd::MirrorGroupSiteStatus &status) {
+  cls::rbd::MirrorGroupSiteStatusOnDisk ondisk_status(status);
+  ondisk_status.mirror_uuid = ""; // mirror_uuid stored in key
+  ondisk_status.up = false;
+  ondisk_status.last_update = ceph_clock_now();
+
+  int r = cls_get_request_origin(hctx, &ondisk_status.origin);
+  ceph_assert(r == 0);
+
+  bufferlist bl;
+  encode(ondisk_status, bl, cls_get_features(hctx));
+
+  r = cls_cxx_map_set_val(hctx,
+                          group_status_global_key(global_group_id,
+                                                  status.mirror_uuid),
+                          &bl);
+  if (r < 0) {
+    CLS_ERR("error setting status for mirrored group, global id '%s', "
+            "site '%s': %s", global_group_id.c_str(),
+            status.mirror_uuid.c_str(),
+            cpp_strerror(r).c_str());
+    return r;
+  }
+  return 0;
+}
+
+int get_remote_group_status_mirror_uuids(cls_method_context_t hctx,
+                                         const std::string& global_group_id,
+                                         std::set<std::string>* mirror_uuids) {
+  std::string filter = group_remote_status_global_key(global_group_id, "");
+  std::string last_read = filter;
+  int max_read = 4; // we don't expect lots of peers
+  bool more = true;
+
+  do {
+    std::set<std::string> keys;
+    int r = cls_cxx_map_get_keys(hctx, last_read, max_read, &keys, &more);
+    if (r < 0) {
+      return r;
+    }
+
+    for (auto& key : keys) {
+      if (!boost::starts_with(key, filter)) {
```

Comment:
I don't get this `if` branch. We're iterating over all OMAP keys in the RBD_MIRRORING object. If we encounter a key that we're not looking for we stop looking?

Reply:
The function is looking for all remote mirror_uuids for  a particular global_ group_id. This translates to looking for keys with the format `gremote_status_global_<global_id>_<remote_mirror_uuid>.
The key to start from  is set to `gremote_status_global_<global_id>_`.
As the keys are returned alphabetically, all the `gremote_status_global_`  keys will be bunched together. The function reads 4 keys (it doesn't expect many peers) starting from the first one to start with `gremote_status_global_<global_id>_`. 
It then checks if they actually begin with the filter. If yes, it strips the mirror_uuid portion from the key and saves it. Otherwise it breaks because there will be no more `gremote_status_global_ ` keys for that global_id.

An example with additional logs added for a single mirror enabled image( instead of group)

``
2024-11-11T15:22:24.671+0530 7f47bce416c0 20 <cls> /workspace/latest/ceph/src/cls/rbd/cls_rbd.cc:5130:  filter = remote_status_global_6a9f8f06-f226-4301-80a5-d232d8a77f96_
2024-11-11T15:22:24.671+0530 7f47bce416c0 20 <cls> /workspace/latest/ceph/src/cls/rbd/cls_rbd.cc:5140:  key = remote_status_global_6a9f8f06-f226-4301-80a5-d232d8a77f96_be894ca6-81d4-4fad-ad35-ea5eed1f8847
2024-11-11T15:22:24.671+0530 7f47bce416c0 20 <cls> /workspace/latest/ceph/src/cls/rbd/cls_rbd.cc:5140:  key = status_global_5d3c9683-5f25-4b85-be5a-79bf06f141d2
2024-11-11T15:22:24.671+0530 7f47bce416c0 20 <cls> /workspace/latest/ceph/src/cls/rbd/cls_rbd.cc:5140: key = status_global_6a9f8f06-f226-4301-80a5-d232d8a77f96
2024-11-11T15:22:24.671+0530 7f47bce416c0 20 <cls> /workspace/latest/ceph/src/cls/rbd/cls_rbd.cc:5143: key = remote_status_global_6a9f8f06-f226-4301-80a5-d232d8a77f96_be894ca6-81d4-4fad-ad35-ea5eed1f8847
2024-11-11T15:22:24.671+0530 7f47bce416c0 20 <cls> /workspace/latest/ceph/src/cls/rbd/cls_rbd.cc:5143: key = status_global_5d3c9683-5f25-4b85-be5a-79bf06f141d2
``


3 keys were returned because there were only 3 keys in rbd_mirroring that came after the filter.


` filter = remote_status_global_6a9f8f06-f226-4301-80a5-d232d8a77f96_`

-  `key = remote_status_global_6a9f8f06-f226-4301-80a5-d232d8a77f96_be894ca6-81d4-4fad-ad35-ea5eed1f8847`
- `key = status_global_5d3c9683-5f25-4b85-be5a-79bf06f141d2`
- `key = status_global_6a9f8f06-f226-4301-80a5-d232d8a77f96`

The first key starts with the filter so the mirror uuid is extracted .
The second key starts with `status_global` instead of `remote_status_global` so there is no point processing any more keys and the code breaks out of the loop.

Reply:
Thanks for the detailed explanation of the optimization. Makes sense